src/provider/provider_devdax_memory.c

```diff
@@ -0,0 +1,677 @@
+                                               &provider->protection);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory protection flags: %u", in_params->protection);
+        return result;
+    }
+
+    /* devdax requires UMF_MEM_MAP_SHARED */
+    result = os_translate_mem_visibility_flag(UMF_MEM_MAP_SHARED,
+                                              &provider->visibility);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory visibility flag: %u", UMF_MEM_MAP_SHARED);
+        return result;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t devdax_initialize(void *params, void **provider) {
+    umf_result_t ret;
+
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_devdax_memory_provider_params_t *in_params =
+        (umf_devdax_memory_provider_params_t *)params;
+
+    if (in_params->path == NULL) {
+        LOG_ERR("devdax path is missing");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (in_params->size == 0) {
+        LOG_ERR("devdax size is 0");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_memory_provider_t *devdax_provider =
+        umf_ba_global_alloc(sizeof(devdax_memory_provider_t));
+    if (!devdax_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    memset(devdax_provider, 0, sizeof(*devdax_provider));
+
+    ret = devdax_translate_params(in_params, devdax_provider);
+    if (ret != UMF_RESULT_SUCCESS) {
+        goto err_free_devdax_provider;
+    }
+
+    devdax_provider->fd = os_devdax_open(in_params->path);
+    if (devdax_provider->fd == -1) {
+        LOG_ERR("cannot open the devdax: %s", in_params->path);
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_free_devdax_provider;
+    }
+
+    devdax_copy_path(in_params->path, devdax_provider->path);
+    devdax_provider->size = in_params->size;
+
+    devdax_provider->base =
+        os_mmap(NULL, devdax_provider->size, devdax_provider->protection,
+                devdax_provider->visibility, devdax_provider->fd, 0);
+    if (devdax_provider->base == NULL) {
+        LOG_PDEBUG("devdax memory mapping failed (path=%s, size=%zu)",
+                   in_params->path, devdax_provider->size);
+        ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_free_devdax_provider;
+    }
+
+    LOG_DEBUG("devdax memory mapped (path=%s, size=%zu)", in_params->path,
+              devdax_provider->size);
+
+    if (util_mutex_init(&devdax_provider->lock) == NULL) {
+        LOG_ERR("lock init failed");
+        ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_free_devdax_provider;
+    }
+
+    devdax_provider->fd_offset_map = critnib_new();
+    if (!devdax_provider->fd_offset_map) {
+        LOG_ERR("creating file descriptor offset map failed");
+        ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_free_devdax_provider;
```

Comment:
`devdax_provider->lock` should probably be destroyed in some separate `err_free_devdax_lock` label.

Reply:
Done

Reply:
@kswiecicki Please verify and resolve.